src/chunkserver/chunkserver_entry.cc

```diff
@@ -0,0 +1,1578 @@
+			safs_pretty_syslog(
+			    LOG_NOTICE,
+			    "Got invalid message in WriteLast state (type:%" PRIu32 ")",
+			    type);
+			state = State::Close;
+			break;
+		}
+	} else if (state == State::WriteForward) {
+		switch (type) {
+		case CLTOCS_WRITE_DATA:
+		case SAU_CLTOCS_WRITE_DATA:
+			writeData(data, type, length);
+			break;
+		case CSTOCL_WRITE_STATUS:
+		case SAU_CSTOCL_WRITE_STATUS:
+			writeStatus(data, type, length);
+			break;
+		case SAU_CLTOCS_WRITE_END:
+			writeEnd(data, length);
+			break;
+		default:
+			safs_pretty_syslog(
+			    LOG_NOTICE,
+			    "Got invalid message in WriteForward state (type:%" PRIu32 ")",
+			    type);
+			state = State::Close;
+			break;
+		}
+	} else if (state == State::WriteFinish) {
+		switch (type) {
+		case CLTOCS_WRITE_DATA:
+		case SAU_CLTOCS_WRITE_DATA:
+		case SAU_CLTOCS_WRITE_END:
+			return;
+		default:
+			safs_pretty_syslog(
+			    LOG_NOTICE,
+			    "Got invalid message in WriteFinish state (type:%" PRIu32 ")",
+			    type);
+			state = State::Close;
+		}
+	} else {
+		safs_pretty_syslog(LOG_NOTICE, "Got invalid message (type:%" PRIu32 ")",
+		                   type);
+		state = State::Close;
+	}
+}
+
+void ChunkserverEntry::checkNextPacket() {
+	TRACETHIS();
+	uint32_t type;
+	uint32_t opSize;
+	const uint8_t *ptr;
+
+	if (state == State::WriteForward) {
+		if (mode == Mode::Data && inputPacket.bytesLeft == 0 &&
+		    fwdBytesLeft == 0) {
+			ptr = headerBuffer;
+			type = get32bit(&ptr);
+			opSize = get32bit(&ptr);
+
+			mode = Mode::Header;
+			inputPacket.bytesLeft = PacketHeader::kSize;
+			inputPacket.startPtr = headerBuffer;
+
+			gotPacket(type, inputPacket.packet.data() + PacketHeader::kSize,
+			          opSize);
+		}
+	} else {
+		if (mode == Mode::Data && inputPacket.bytesLeft == 0) {
+			ptr = headerBuffer;
+			type = get32bit(&ptr);
+			opSize = get32bit(&ptr);
+
+			mode = Mode::Header;
+			inputPacket.bytesLeft = PacketHeader::kSize;
+			inputPacket.startPtr = headerBuffer;
+			gotPacket(type, inputPacket.packet.data(), opSize);
+		}
+	}
+}
+
+void ChunkserverEntry::fwdConnected() {
+	TRACETHIS();
+	int status = tcpgetstatus(fwdSocket);
+	if (status) {
+		safs_silent_errlog(LOG_WARNING, "connection failed, error");
+		fwdError();
+		return;
+	}
+	tcpnodelay(fwdSocket);
+	state = State::WriteInit;
+}
+
+void ChunkserverEntry::fwdRead() {
+	TRACETHIS();
+	int32_t bytesRead;
+	uint32_t type;
+	uint32_t opSize;
+	const uint8_t *ptr;
+
+	if (fwdMode == Mode::Header) {
+		bytesRead =
+		    read(fwdSocket, fwdInputPacket.startPtr, fwdInputPacket.bytesLeft);
+		if (bytesRead == 0) {
+			fwdError();
+			return;
+		}
+		if (bytesRead < 0) {
+			if (errno != EAGAIN) {
+				safs_silent_errlog(LOG_NOTICE, "(fwdread) read error");
+				fwdError();
+			}
+			return;
+		}
+		stats_bytesin += bytesRead;
+		fwdInputPacket.startPtr += bytesRead;
+		fwdInputPacket.bytesLeft -= bytesRead;
+		if (fwdInputPacket.bytesLeft > 0) {
+			return;
+		}
+		ptr = fwdHeaderBuffer + 4;
```

Comment:
Please consider specifying a variable naming the 4 value for making clear why is this value used